### ISSUE TITLE
Calling getcurpos() changes curswant in visual block mode

### ISSUE THREAD

**Issue by w0rp** [vim/vim#4069]
Here is a strange bug that's easy to repeat. Start Vim with no plugins like `vim -u NONE`, and add two lines to a buffer like so.

```
one
two
```

Type `<C-V>jI123<space><esc>j` to end up with lines like so. (Inserting "123 " before both lines in Visual Block mode.)

```
123 one
123 two
```

Your cursor should be on the first column of the second line. Now start again with the lines as before, and add `:autocmd InsertLeave * call getcurpos()` before editing the text, inserting the text exactly as before. Your cursor will end up on the space between "123" and "two" instead.

In other words, calling `getcurpos()` changes the value of curswant when called from an `autocmd` command while editing in visual block mode.

**Comment by andymass** [comment]
This is undocumented and probably unintended, but it is necessary sometimes to use this side-effect:

https://github.com/natebosch/vim-lsc/blob/f7adfe36af3afbd557c885819b6218a256f375ea/autoload/lsc/reference.vim#L112

Using `getcurpos()` was the only way to get this code to work.  No combination of `cursor()` or `setpos()` would do it.  So, if `getcurpos()` is changed, an alternative way be provided to set `curswant`?

Alternatively, can that other bug be fixed before altering `getcurpos()`?  Specifically, if the cursor is being set inside channel callbacks, and `curswant` is wrong unless you call `getcurpos()`.

**Comment by chrisbra** [comment]
>  an alternative way be provided to set curswant

I have not looked into the bug yet, but you can also set the `curswant` property with `winsaveview()` (by feeding it a dict with only the `curswant` key).

**Comment by andymass** [comment]
@chrisbra , I tried this

    call cursor(a:line, a:character)
    call winrestview({'curswant': virtcol('.')-1})

However in my use-case (inside a particular callback), this does not trigger `CursorMoved`.

[edit] ~~If I then manually do `:call getcurpos()`  afterwards, I get a `CursorMoved`, which is very bizarre behavior.~~

In my original tests there seemed to be something wrong with setting the cursor position which is "fixed" reliably by ` call getcurpos()`.  But it's possible I was just confused by the virtual column (`getcurpos()` handles this without having to compute the curswant virtual column).

**Comment by andymass** [comment]
@chrisbra Sorry, I spoke too soon.  The CursorMoved is being triggered by the entering / leaving the command line, and is also triggered by a redraw (`ctrl-l`).  I don't know if this actually has anything to do with `curswant`, but the behavior is still buggy somehow.

**Comment by brammool** [comment]
CursorMoved is only triggered from the main loop, not from a callback.
I think we should keep it this way, triggering an autocommand from
further down is likely to cause trouble.

-- 
Despite the cost of living, have you noticed how it remains so popular?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by w0rp** [comment]
I don't care about triggering `CursorMoved`. My issue remains. Calling `getcurpos()` changes the value of `curswant`, which means that any plugins which check the cursor position when you modify lines in visual block mode will change the cursor position. It doesn't make sense for the cursor position to be modified when you read the cursor position.

**Comment by chrisbra** [comment]
@w0rp: Just to make clear: Patch 8.1.0998 has fixed it, right?

**Comment by w0rp** [comment]
Ah sorry, yes that does fix it. Thanks! I missed the patch.
